### PR TITLE
Improve `postprocessing` performance by adjusting its logic(around 3.5x faster)

### PR DESCRIPTION
**Pull request by danroux**
Inside client#postprocessing https://github.com/Barzahlen/grac/blob/master/lib/grac/client.rb#L177 we go through each data key and compare it to each and every key present on @options[:postprocessing], even if there was already a successful match to pick the lambda to run, we continue making comparisons which is useless, since no other will match and because of this wastes time.

I propose that we stop looping on @options[:postprocessing] on 1st match, pick the lambda and continue processing data.

In my initial tests i've seen this makes the method around 3.5-4 times faster.

### CODE REVIEW

File: lib/grac/client.rb [hunk 1]
```diff
@@ -29,6 +29,9 @@ def initialize(uri, options = {})
       [:params, :headers, :postprocessing, :middleware].each do |k|
         @options[k].freeze
       end
+      @patterns = @options[:postprocessing].keys.collect do |regex|
```
Comment (tobischo) [inline lib/grac/client.rb]: By doing this in initialize `@patterns` is built every time a client is initialized.
Since grac makes heavy usage of method chaining, maybe we should instead keep initializing the `Regexp` objects in `postprocessing` but utilizing memoization

File: lib/grac/client.rb [hunk 2]
```diff
@@ -172,16 +175,14 @@ def postprocessing(data, processing = nil)
       return data if @options[:postprocessing].nil? || @options[:postprocessing].empty?
 
       if data.kind_of?(Hash)
-        data.each do |key, value|
+        data.keys.reverse.each do |key|
```
Comment (tobischo) [inline lib/grac/client.rb]: Could you explain the necessity of `.reverse` here?

On that note: Why did you move to looping over the keys instead of keys and values?

File: lib/grac/client.rb [hunk 3]
```diff
@@ -25,13 +25,24 @@ def initialize(uri, options = {})
...
-      @options.freeze
+
+      if options[:postprocessing]
+        options[:postprocessing]
+          .each_with_object(patterns = { postprocessing: {} }) do |(key, value), obj|
+          if !key.kind_of?(Regexp)
+            obj[:postprocessing][Regexp.new(key)] = value
+          end
+        end
+        options.merge!(patterns)
```
Comment (tobischo) [inline lib/grac/client.rb]: Looks good, although I am wondering:
* `patterns` will only contain `Regexp` + `transformation lambda`: Wouldn't `patterns = {}` and `obj[Regexp.new(key)] = value` be sufficient?
* Could it be more efficient to have an else condition instead of a merge later on:
```ruby
if !key.kind_of?(Regexp)
  obj[:postprocessing][Regexp.new(key)] = value
else
  obj[:postprocessing][key] = value
end
```
* If you'd skip the merge part (only values with `Regexp` keys should make it into patterns then) you could probably skip the `delete_if` part
* Does it make sense to initialize `@options` with `:postprocessing => options[:postprocessing] || {},`? if `:postprocessing` is not null, it will be overwritten, otherwise it will be `{}`, so maybe `:postprocessing => {},` would be more fitting now to cover the default case
* Did you remove `@options.freeze` on purpose or by accident?

Comment (tobischo) [reply lib/grac/client.rb]: Suggestion:
```ruby
@options = {
  :connecttimeout => options[:connecttimeout] || 0.1,
  :timeout        => options[:timeout]        || 15,
  :params         => options[:params]         || {},
  :headers        => {
    "User-Agent"   => "Grac v#{Grac::VERSION}",
    "Content-Type" => "application/json;charset=utf-8"
  }.merge(options[:headers] || {}),
  :postprocessing => {},
  :middleware     => options[:middleware]     || []
}

if options[:postprocessing]

  options[:postprocessing]
    .each_with_object(postprocessing = {}) do |(key, value), obj|
    if key.kind_of?(Regexp)
      obj[key] = value
    else
      obj[Regexp.new(key)] = value
    end
  end
  
  @options[:postprocessing] = postprocessing
end
@options.freeze
```

Comment (danroux) [reply lib/grac/client.rb]: I agree with your suggestions and i just made the corresponding changes. To address some of your points:

> patterns will only contain Regexp + transformation lambda: Wouldn't patterns = {} and obj[Regexp.new(key)] = value be sufficient?

Now that we are going to do `@options[:postprocessing] = postprocessing` i think it makes more sense to initialize an empty hash.

> Did you remove @options.freeze on purpose or by accident?

Yes 😊. Fixed now.

I also renamed some variables: `key -> pattern`, `value -> transformation` and inside `#postprocessing` i removed `patterns` and just take and iterate on `@options[:postprocessing].keys` directly.
